### PR TITLE
Fix transit stuff

### PR DESCRIPTION
**Pull request by Maxopoly**
Players stay in transit until another shards sends a message indicating that the player arrived or after a 30 second time out.

### CODE REVIEW

**Comment by Maxopoly** [comment]
Oh and portals have a 10 second cooldown now

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java [hunk 1]
```diff
@@ -150,20 +134,23 @@ private void addPlayerToTransit(final UUID uuid, final String server){
...
 			}
 			
-		}, 20);
+		}, 600); //30 seconds timeout
 	}
+	
+	public void notifySuccessfullTransfer(UUID player) {
+		if (isPlayerInTransit(player)) {
+			//TODO Readd PlayerEnsuredToTransitEvent here
+			transit.remove(player);
```
Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java]: This doesn't cancel the timeout event though

Comment (Maxopoly) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java]: Doesnt really matter, because the timeout will first check, whether the player is still in transit. If he isnt, it will do nothing.

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/external/MercuryManager.java [hunk 2]
```diff
@@ -65,54 +65,58 @@ public void teleportPlayer(UUID uuid, Portal p, Object ... data) {
...
 		MercuryAPI.sendGlobalMessage("portal|connect|"+main.getName()+"|"+con.getName(), "BetterShards");
 	}
 	
-	public void removePortalJoin(Portal main) {
+	public static void removePortalJoin(Portal main) {
 		MercuryAPI.sendGlobalMessage("portal|remove|"+main.getName(), "BetterShards");
 	}
 	
-	public void sendPlayerCount(int count) {
+	public static void sendPlayerCount(int count) {
```
Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/external/MercuryManager.java:115]: Why the change to all static?

Comment (Maxopoly) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/external/MercuryManager.java:115]: Because it is static. Doesnt depend on an instance of the class and is just an API, so passing around an instance is pointless and it might as well be static. Not really related to the other changes, just something I stumbled across.

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java (hunk 1 above)
Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java]: Ah missed that line, carry on

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java [hunk 3]
```diff
@@ -150,20 +134,23 @@ private void addPlayerToTransit(final UUID uuid, final String server){
...
-				transit.remove(uuid);
-				Player player = Bukkit.getPlayer(uuid);
-				if (player != null && !player.isOnline()) {
+				if (isPlayerInTransit(uuid)) {
+					info(uuid + " failed to transit, was removed by timeout");
 					PlayerFailedToTransitEvent event = new PlayerFailedToTransitEvent(uuid, server);
 					Bukkit.getPluginManager().callEvent(event);
 				}
-				else {
-					PlayerEnsuredToTransitEvent event = new PlayerEnsuredToTransitEvent(uuid, server);
```
Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java:160]: Any particular reason to remove this event?

Comment (Maxopoly) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java:160]: It has to be called based on the callback of the other server and right now it doesnt keep track of which server the player was sent to after he's gone, so I cant call the event. I could turn the list into a map though which uses the server the player was sent to as value.

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java [hunk 4]
```diff
@@ -145,30 +130,34 @@ public static BetterShardsPlugin getInstance(){
...
-				}
 			}
 			
-		}, 20);
+		}, 600); //30 seconds timeout
 	}
+	
+	public void notifySuccessfullTransfer(UUID player) {
+		if (isPlayerInTransit(player)) {
+			PlayerEnsuredToTransitEvent e = new PlayerEnsuredToTransitEvent(player, transit.get(player));
```
Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java]: Perfect fix

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java [hunk 5]
```diff
@@ -69,7 +68,7 @@
 	private Map<Player, Grid> grids = new HashMap<Player, Grid>();
 	private Map<UUID, BedLocation> beds = new HashMap<UUID, BedLocation>();
 	
-	private List<UUID> transit = new ArrayList<UUID>();
+	private Map<UUID, String> transit = new TreeMap<UUID, String>();
```
Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java]: TreeMap isn't synchronized; if you need the sorted sequencing use `SortedMap m = Collections.synchronizedSortedMap(new TreeMap(...));` otherwise just use `ConcurrentHashMap`

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/PortalsManager.java [hunk 6]
```diff
@@ -21,8 +22,9 @@
 
 	private DatabaseManager db = BetterShardsPlugin.getInstance().getDatabaseManager();
 	private Map<String, Portal> portals;
-	private List<Player> arrivedPlayers = new ArrayList<Player>();
+	private Map<UUID, Long> arrivedPlayers = new TreeMap<UUID,Long>();
```
Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/PortalsManager.java]: Same issue here; if you need sorting used the externally synchronized SortedMap interfaces or if sort isn't needed use ConcurrentHashMap

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java (hunk 5 above)
Comment (Maxopoly) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java]: All access to this map is sync though?

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/PortalsManager.java [hunk 7]
```diff
@@ -105,33 +107,38 @@ private void removeTeleportedPlayers(){
...
-						List<Player> toRemove = new ArrayList<Player>();
-						for (Player p: arrivedPlayers){
-							if (!p.isOnline()) {
-								toRemove.add(p);
+						List<UUID> toRemove = new LinkedList<UUID>();
+						long currTime = System.currentTimeMillis();
+						for (Entry <UUID, Long> entry: arrivedPlayers.entrySet()){
+							Player p = Bukkit.getPlayer(entry.getKey());
+							if (p == null) {
 								continue;
```
Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/PortalsManager.java:114]: If a player logs off and never arrives this will just keep the player in the list forever.

On player not found probably just remove them? Not sure.

Comment (Maxopoly) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/PortalsManager.java:114]: Keep in mind that players are only added to this list in the join event. Being in this list guarantees that they arrived and either are still standing in a portal or their timer isnt over yet. Keeping them after logoff is intended, because it means they wont be bounced back immediately once they login, but instead can walk out of the portal and only are sent back if they reenter it.

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/listeners/MercuryListener.java [hunk 8]
```diff
@@ -162,6 +162,10 @@ else if (content[0].equals("info")) {
 				}
 			}
 		}
+		else if (content [0].equals("arrival")) {
+			UUID uuid = UUID.fromString(content [1]);
+			plugin.notifySuccessfullTransfer(uuid);
```
Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/listeners/MercuryListener.java]: No, access isn't all synchronous to `transit`.

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/PortalsManager.java (hunk 7 above)
Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/PortalsManager.java:114]: So that kind of caching would persist until restart; useful but good to note its limits.

So teleport actually lands you inside the other side's portal? Not outside of it?

Comment (Maxopoly) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/PortalsManager.java:114]: Assuming it's a 2 way portal, you will always land inside a portal, yes.

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java (hunk 5 above)
Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java]: MercuryListener triggers a function that modifies it; that mercury listener is async.

Comment (Maxopoly) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/BetterShardsPlugin.java]: Yeah you are right, already pushed a fix

**Comment by ProgrammerDan** [comment]
Ok so the only thing that seems to be missing is blocking saves while intransit.

Also, will the target server know the player is intransit? Ideally we'd want to cancel events and saves on the target server until the player is fully arrived.

**Comment by Maxopoly** [comment]
Right now the target server doesn't know the difference between a normal player login and a player arriving because of Bungee. We could do something with Mercury, but it's hard to guarantee anything.

**Comment by Maxopoly** [comment]
Do you mean that we should add an inTransit equivalent for arrival that prevents anything until player data is guaranteed to be loaded?

**Comment by ProgrammerDan** [comment]
That's fine, mostly I'm just after what I described in the channel:

while intransfer prevent save, 'cept for the save which is triggered by the transfer event

similarly on join, prevent save until player is fully joined and the message releasing intransfer is sent.

**Comment by Maxopoly** [comment]
Ok, I'll refrain from making the main plugin class worse and split that and the original in transit thing out in a separate class

**Comment by ProgrammerDan** [comment]
basically Intransit should tell the target server to expect the person; then they can be intransit on both sides.

This doesn't conflict with normal first-joins as the player will simply join as normal, with no interference from intransit b/c they won't be on that list.

**Comment by Maxopoly** [comment]
Can we guarantee that the Mercury message arrives before the player arrives via Bungee?

**Comment by ProgrammerDan** [comment]
not yet but it doesn't matter; if the message arrives late we'll just be "unprotected" but it shouldn't interfere with normal login as the arriving server will assume they are a user logging straight in to them instead of a transfer. The bit we're discussing is just extra protection.

In terms of order of operations however, the intransit alert should be sent prior to anything else in the transfer sequence; rabbit delivers pretty darn fast so there's a very good chance it will happen first.

**Comment by Maxopoly** [comment]
No, that was missing. I added it.

**Comment by Maxopoly** [comment]
So when should I be removing players from arrivalTransit and send a message to the origin shard to clean up it's exitTransit as well?

Right now it's in the join listener, should it be at the end of the asyncprelogin listener?

Aside from that is there anything else that should go here before it's merged?

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/listeners/BetterShardsListener.java [hunk 9]
```diff
@@ -370,25 +375,25 @@ public void playerSleepInBed(PlayerInteractEvent event) {
...
+		if(BetterShardsPlugin.getTransitManager().isPlayerInTransit(player.getUniqueId())) {
 			event.setCancelled(true);
 		}
 	}
 	
-	@EventHandler(priority = EventPriority.HIGHEST)
+	@EventHandler(priority = EventPriority.LOWEST)
 	public void playerDamageEvent(EntityDamageEvent event) {
-		if (!(event.getEntity() instanceof Player))
+		if (!(event.getEntity().getType() == EntityType.PLAYER))
```
Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/listeners/BetterShardsListener.java]: So the only downside here is with this change, line 390 is an unchecked cast, whereas before it was checked.

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/manager/TransitManager.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+			exitTransit.remove(player);
+		}
+	}
+	
+	/**
+	 * Called when a player's data is fully loaded
+	 * 
+	 * @param player UUID of the transferred player
+	 */
+	public void notifySuccessfullArrival(UUID player) {
```
Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/manager/TransitManager.java:95]: Shouldn't this be checking isPlayerInArrivalTransit() and using arrivalTransit?

And if it throws an event, perhaps a PlayerEnsuredFromTransitEvent ?

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/listeners/BetterShardsListener.java (hunk 9 above)
Comment (Maxopoly) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/listeners/BetterShardsListener.java]: Well but anything with Player as EntityType will also be a player. Is this just a compiler warning thing? How bad is using instanceof actually?

Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/listeners/BetterShardsListener.java]: Using instanceof is never bad.

The Type isn't strictly related to the class (cast type) it's just an internal Bukkit thing fwiw. So if you know it needs to fulfill the Player interface, check `instanceof`. If you're okay with dealing with it as an Entity but also would like to note its a player, use EntityType comparison.

**Comment by Maxopoly** [comment]
Ok fixed those. Thoughts on what I asked above?

File: BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/misc/CustomWorldNBTStorage.java [hunk 11]
```diff
@@ -54,7 +69,12 @@ public void save(EntityHuman entityhuman) {
 			NBTCompressedStreamTools.a(nbttagcompound, output);
 
 			// Now to run our custom mysql code
-			db.savePlayerData(uuid, output, getInvIdentifier(uuid), sect.get(uuid));
+			if (!BetterShardsPlugin.getTransitManager().isPlayerInTransit(uuid)) {
+				db.savePlayerData(uuid, output, getInvIdentifier(uuid), sect.get(uuid));
+			}
+			else {
+				logger.log(Level.WARNING, "Didn't save data for {0}, because he was in transit", entityhuman.getName());
```
Comment (ProgrammerDan) [inline BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/misc/CustomWorldNBTStorage.java:76]: So a gotcha here; `'` is parsed as "read everything that follows as a string literal and don't interpret". That's why if you look at my other log lines I never use apostraphes or contractions. So easy fix; just "Did not save..." and that's it.

**Comment by ProgrammerDan** [comment]
So I _think_ we should remove them from origin server's exit transit when on the target server the player has finished join (Join MONITOR would be a fine place for it). There's no issue "keeping" the player in the exit queue, only reason to not just let them time out is in case the player wants to go back shortly after 10 sec timeout. 

In fact you could just spawn an event to run 1-2s after join, maybe that would be best.
